cvat/apps/engine/cache.py

```diff
@@ -54,74 +61,225 @@
     ZipChunkWriter,
     ZipCompressedChunkWriter,
 )
-from cvat.apps.engine.utils import load_image, md5_hash
+from cvat.apps.engine.rq_job_handler import RQJobMetaField
+from cvat.apps.engine.utils import get_rq_lock_for_job, load_image, md5_hash
 from utils.dataset_manifest import ImageManifestManager
 
 slogger = ServerLogManager(__name__)
 
 
 DataWithMime = Tuple[io.BytesIO, str]
-_CacheItem = Tuple[io.BytesIO, str, int]
+_CacheItem = Tuple[io.BytesIO, str, int, Union[datetime, None]]
 
 
-class MediaCache:
-    def __init__(self) -> None:
-        self._cache = caches["media"]
+class CvatCacheTimestampMismatchError(Exception):
+    pass
+
+
+def _is_run_inside_rq() -> bool:
+    return rq.get_current_job() is not None
+
+
+def _convert_args_for_callback(func_args: list[Any]) -> list[Any]:
+    result = []
+    for func_arg in func_args:
+        if _is_run_inside_rq():
+            result.append(func_arg)
+        else:
+            if isinstance(
+                func_arg,
+                (models.Task, models.Segment, models.Job, models.CloudStorage, models.Data),
+            ):
+                result.append(func_arg.id)
+            elif isinstance(func_arg, list):
+                result.append(_convert_args_for_callback(func_arg))
+            else:
+                result.append(func_arg)
+
+    return result
+
+
+@attrs.define
+class Callback:
+    _callable: Callable[..., DataWithMime]
+    _args: list[Any] = attrs.field(converter=_convert_args_for_callback, factory=list)
+    _kwargs: dict[str, Any] = attrs.Factory(dict)
```

Comment:
It seems like there must be a check that only specific types (only primitives?) are passed in kwargs.